src/main/cpp/checkpoint/CheckPoint.cpp

```diff
@@ -0,0 +1,675 @@
+	hid_t newType = H5Dget_type(clusterID);
+
+	hsize_t dims;
+	H5Sget_simple_extent_dims(dspace, &dims, nullptr);
+
+	hsize_t start = 0;
+
+	hsize_t count = 1;
+
+	hid_t subspace = H5Dget_space(clusterID);
+	H5Sselect_hyperslab(subspace, H5S_SELECT_SET, &start, nullptr, &count, nullptr);
+	clusterData data;
+
+	H5Dread(clusterID, newType, H5S_ALL, subspace, H5P_DEFAULT, &data);
+
+	H5Sclose(subspace);
+
+	Cluster* CurrentCluster = new Cluster(data.ID, i);
+
+	for (hsize_t j = 1; j < dims; j++) {
+		hsize_t start = j;
+		hsize_t count = 1;
+
+		hid_t subspace = H5Screate_simple(1, &count, nullptr);
+
+		H5Sselect_hyperslab(dspace, H5S_SELECT_SET, &start, nullptr, &count, nullptr);
+
+		clusterData data;
+
+		H5Dread(clusterID, newType, subspace, dspace,H5P_DEFAULT, &data);
+
+		H5Sclose(subspace);
+
+		if (data.ID != CurrentCluster->GetId()) {
+			clusters.emplace_back(*CurrentCluster);
+			CurrentCluster = new Cluster(data.ID, i);
+			continue;
+		}
+
+		unsigned int idPersonToAdd = data.PersonID;
+
+		result.serial_for([this, &idPersonToAdd, &CurrentCluster](const Person& p, unsigned int) {
+			if (p.GetId() == idPersonToAdd) {
+				CurrentCluster->AddPerson(p);
+			}
+		});
+	}
+
+	clusters.emplace_back(*CurrentCluster);
+	H5Tclose(newType);
+	H5Dclose(clusterID);
+	H5Sclose(dspace);
+}
+
+SingleSimulationConfig CheckPoint::LoadSingleConfig()
+{
+	htri_t exist = H5Lexists(m_file, "Config", H5P_DEFAULT);
+	if (exist <= 0) {
+		FATAL_ERROR("Invalid file");
+	}
+
+	SingleSimulationConfig result;
+	auto comCon = std::make_shared<CommonSimulationConfig>();
+	auto logCon = std::make_shared<LogConfig>();
+	result.common_config = comCon;
+	result.log_config = logCon;
+
+	WriteDSetFile("tmp_matrix.xml", "Config/disease");
+	result.common_config->contact_matrix_file_name = "tmp_matrix.xml";
+
+	WriteDSetFile("tmp_disease.xml", "Config/contact");
+	result.common_config->disease_config_file_name = "tmp_disease.xml";
+
+	hid_t group = H5Gopen2(m_file, "Config", H5P_DEFAULT);
+
+	hid_t attr = H5Aopen(group, "bools", H5P_DEFAULT);
+	hbool_t bools[2];
+	H5Aread(attr, H5T_NATIVE_HBOOL, bools);
+	H5Aclose(attr);
+
+	result.common_config->track_index_case = bools[0];
+	result.log_config->generate_person_file = bools[1];
+
+	attr = H5Aopen(group, "doubles", H5P_DEFAULT);
+	double doubles[3];
+	H5Aread(attr, H5T_NATIVE_DOUBLE, doubles);
+	H5Aclose(attr);
+
+	result.common_config->r0 = doubles[0];
+	result.common_config->seeding_rate = doubles[1];
+	result.common_config->immunity_rate = doubles[2];
+
+	attr = H5Aopen(group, "uints", H5P_DEFAULT);
+	unsigned int uints[3];
+	H5Aread(attr, H5T_NATIVE_UINT, uints);
+	H5Aclose(attr);
+
+	result.common_config->rng_seed = uints[0];
+	result.common_config->number_of_days = uints[1];
+	result.common_config->number_of_survey_participants = uints[2];
+	result.log_config->log_level = (LogMode)uints[3];
+	unsigned int id = uints[4];
+
+	attr = H5Aopen(group, "prefix", H5P_DEFAULT);
+
+	H5A_info_t* info = new H5A_info_t();
```

Comment:
Rewriting this line as `auto info = std::make_unique<H5A_info_t>();` and deleting the `delete info;` statement below will save you a line of code and make your code more robust to future changes.